### PR TITLE
Add room version upgrades

### PR DESCRIPTION
**Pull request by turt2live**
Rendered: see 'docs' status check
Proposal: https://github.com/matrix-org/matrix-doc/issues/1501

### CODE REVIEW

**Comment by turt2live** [comment]
I've asked for the review of the whole team where, but would appreciate particular review from @richvdh to ensure accuracy to the proposal and one other member for general review.

File: specification/modules/room_upgrades.rst [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+{{m_room_tombstone_event}}
+
+Client behaviour
+----------------
+
+Clients which understand ``m.room.tombstone`` events MUST:
+
+* Hide the old room from the user's list of rooms. Permalinks, backlinks, etc should
+  still be accessible to this room, however.
+* At the very bottom of the old room's timeline/message view, display a message which
```
Comment (ara4n) [inline specification/modules/room_upgrades.rst]: do we really want to mandate that clients have to split the room into two separate views with permalinks between them? if the client is smart enough, shouldn't it stitch the two views together into a single timeline?

Comment (turt2live) [reply specification/modules/room_upgrades.rst]: I was trying to suggest something like that by saying "message view", although I managed to drown it out with the rest of the text :(

I'm not sure if we need the strict UX requirements here. @richvdh what do you think?

File: event-schemas/schema/m.room.create [hunk 2]
```diff
@@ -14,6 +14,18 @@ properties:
...
         type: string
+      predecessor:
+        description: A reference to the room this room replaces, if the previous room was upgraded.
+        type: object
+        title: Previous Room
+        properties:
+          room_id:
+            type: string
+            description: The ID of the old room.
+          event_id:
```
Comment (richvdh) [inline event-schemas/schema/m.room.create]: we should probably decide if we're going to do https://github.com/matrix-org/matrix-doc/issues/1774 before we bake this into the spec.

File: specification/modules/room_upgrades.rst (hunk 1 above)
Comment (richvdh) [reply specification/modules/room_upgrades.rst]: It's much too strongly worded at present; it should be up to the client to figure out what a sensible UX should be.

I don't mind if we want to present this as an example of a plausible UX, but saying that clients MUST do all this stuff is too strong.

File: event-schemas/schema/m.room.create (hunk 2 above)
Comment (turt2live) [reply event-schemas/schema/m.room.create]: When I opened 1774 the intention was that it'd be fixed a long ways down the road, maybe. Not anytime soon.

**Comment by anoadragon453** [comment]
Note: we should be somewhat vague about exactly what needs to be copied over, as more and more will be necessary in the future, and we don't want to keep updating the doc each time lest it get out of sync.

**Comment by turt2live** [comment]
The UX requirements should be a lot less imposing now. I've also clarified what the server should be transferring while still being open-ended/vague enough to permit a wide range of things.

@richvdh @ara4n @anoadragon453 Please take another look.

File: specification/modules/room_upgrades.rst [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+------
+
+{{m_room_tombstone_event}}
+
+Client behaviour
+----------------
+
+Clients which understand ``m.room.tombstone`` events and the ``predecessor`` field on
+``m.room.create`` events should communicate to the user that the room was upgraded.
+One way of accomplishing this would be to hide the old room from the user's room list
```
Comment (richvdh) [inline specification/modules/room_upgrades.rst]: s/to hide/hiding/ for consistency with the rest of the sentence.

**Comment by turt2live** [comment]
@ara4n just want to double check that the improved wording is in line with what you would expect.

**Pull request merged**